src/ClientData/CallstackDataTest.cpp

```diff
@@ -200,4 +200,124 @@ TEST(CallstackData, FilterCallstackEventsBasedOnMajorityStartExcludesFunctionToS
+  callstack_list.clear();
+  callstack_data.ForEachCallstackEventOfTidInTimeRangeDiscretized(
+      kTid, 0, std::numeric_limits<uint64_t>::max(), kResolution, insert_callstack);
+  EXPECT_THAT(callstack_list,
+              testing::Pointwise(CallstackEventEq(), std::vector<CallstackEvent>{event1}));
+}
+
+TEST(CallstackData, ForEachCallstackEventInTimeRangeDiscretized) {
+  CallstackData callstack_data;
+
+  const uint64_t cs_id = 12;
+  CallstackInfo cs{{0x11, 0x10}, CallstackType::kComplete};
+  callstack_data.AddUniqueCallstack(cs_id, std::move(cs));
+
+  constexpr uint32_t kTid = 42;
+  constexpr uint32_t kAnotherTid = 43;
+
+  const uint64_t time1 = 142;
+  CallstackEvent event1{time1, cs_id, kTid};
+  callstack_data.AddCallstackEvent(event1);
+
+  const uint64_t time2 = 242;
+  CallstackEvent event2{time2, cs_id, kTid};
+  callstack_data.AddCallstackEvent(event2);
+
+  const uint64_t time3 = 342;
+  CallstackEvent event3{time3, cs_id, kAnotherTid};
+  callstack_data.AddCallstackEvent(event3);
+
+  const uint64_t time4 = 442;
+  CallstackEvent event4{time4, cs_id, kTid};
+  callstack_data.AddCallstackEvent(event4);
+
+  std::vector<CallstackEvent> callstack_list;
+  auto insert_callstack = [&](const CallstackEvent& event) { callstack_list.push_back(event); };
+  const uint32_t kResolution = 2000;
+  callstack_data.ForEachCallstackEventInTimeRangeDiscretized(0, 1000, kResolution,
+                                                             insert_callstack);
+  EXPECT_THAT(callstack_list,
+              testing::Pointwise(CallstackEventEq(),
+                                 std::vector<CallstackEvent>{event1, event2, event3, event4}));
+
+  callstack_list.clear();
+  callstack_data.ForEachCallstackEventInTimeRangeDiscretized(0, event2.timestamp_ns() + 1,
+                                                             kResolution, insert_callstack);
+  EXPECT_THAT(callstack_list,
+              testing::Pointwise(CallstackEventEq(), std::vector<CallstackEvent>{event1, event2}));
+
+  // With one pixel on the screen we should only see one event. It should be the first of some of
+  // the threads.
+  callstack_list.clear();
+  callstack_data.ForEachCallstackEventInTimeRangeDiscretized(0, 1000, /*resolution=*/1,
+                                                             insert_callstack);
+  EXPECT_EQ(callstack_list.size(), 1);
+  EXPECT_THAT(callstack_list[0], testing::AnyOf(event1, event3));
+
+  // When max_timestamp is std::numeric_limits<uint64_t>::max(), each callstack should be draw in
+  // the first pixel, and therefore only one will be visible. It should be the first of some of the
+  // threads.
+  callstack_list.clear();
+  callstack_data.ForEachCallstackEventInTimeRangeDiscretized(
+      0, std::numeric_limits<uint64_t>::max(), kResolution, insert_callstack);
+  EXPECT_EQ(callstack_list.size(), 1);
+  EXPECT_THAT(callstack_list[0], testing::AnyOf(event1, event3));
```

Comment:
Just like in the test above, I'd add a function. But it would have to take a matcher, not just a `vector<CallstackEvent>`.

Reply:
I made a big refactor, PTAL.